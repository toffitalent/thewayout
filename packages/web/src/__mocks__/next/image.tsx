import * as React from 'react';

// eslint-disable-next-line react/jsx-no-useless-fragment
const mock = (props: { children: React.ReactElement }): React.ReactElement => <>{props.children}</>;

export default mock;
