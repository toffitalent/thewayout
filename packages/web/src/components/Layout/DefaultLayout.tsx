import { SkipNavContent, SkipNavLink } from '@disruptive-labs/ui';
import React from 'react';
import { Header } from './Header';

export interface DefaultLayoutProps {
  header?: React.ComponentType;
  children?: React.ReactNode;
}

export const DefaultLayout: React.FC<DefaultLayoutProps> = ({
  children,
  header: HeaderComponent = Header,
}) => (
  <>
    <SkipNavLink>Skip to content</SkipNavLink>
    <HeaderComponent />
    <SkipNavContent />
    <main>{children}</main>
  </>
);
