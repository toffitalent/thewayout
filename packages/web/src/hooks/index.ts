export * from './useAppAuth';
export * from './useAppDispatch';
export * from './useAppSelector';
export * from './useDashboardRedirection';
export * from './useErrorIndicator';
export * from './useFormattedPhoneNumber';
export * from './useIndicators';
export * from './useMediaQuery';
export * from './usePrevious';
export * from './useTimeout';
