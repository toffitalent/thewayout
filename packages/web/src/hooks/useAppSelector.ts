import { TypedUseSelectorHook, useSelector } from 'react-redux';
import type { State } from '@app/store';

export const useAppSelector: TypedUseSelectorHook<State> = useSelector;
