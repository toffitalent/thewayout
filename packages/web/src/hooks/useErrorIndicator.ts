import { Toast } from '@disruptive-labs/ui';
import { useCallback } from 'react';
import { ApiError } from '@app/api';
import { isSerializedApiError, SerializedApiError } from '@app/utils';

const getDefaultMessageByErrorCode = (error: ApiError) => {
  switch (error.code) {
    case 'RequestThrottled':
      return 'Too many requests. Please try again in a few minutes.';
    default:
      return null;
  }
};

const getErrorMessage = (
  error: ApiError | SerializedApiError,
  errorCodeMap: Record<string, string>,
  defaultMessage: string | null,
) => {
  switch (error.type) {
    case ApiError.Types.Client:
      return errorCodeMap?.[error.code] ?? getDefaultMessageByErrorCode(error) ?? defaultMessage;
    case ApiError.Types.Network:
      return 'No connection';
    default:
      return defaultMessage;
  }
};

export const useErrorIndicator = (
  errorCodeMap?: Record<string, string>,
  defaultMessage: string | null = 'Something went wrong',
) =>
  useCallback(
    (error: any, shouldRethrowError = true) => {
      const text =
        error instanceof ApiError || isSerializedApiError(error)
          ? getErrorMessage(error, errorCodeMap ?? {}, defaultMessage)
          : defaultMessage;

      if (text) Toast.show({ status: 'danger', text });

      if (shouldRethrowError) throw error;
    },
    [defaultMessage, errorCodeMap],
  );
