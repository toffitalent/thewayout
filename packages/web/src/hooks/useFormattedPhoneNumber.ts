import type { AsYouType, CountryCode, PhoneNumber } from 'libphonenumber-js';
import { useEffect, useState } from 'react';

type ParsePhoneNumber = (text: string, defaultCountry?: CountryCode | undefined) => PhoneNumber;

const loadModule = async () => {
  const { AsYouType, parsePhoneNumber } = await import('libphonenumber-js');

  return { AsYouType, parsePhoneNumber };
};

export const useFormattedPhoneNumber = () => {
  const [formatNumber, setFormatNumber] = useState<AsYouType>();
  const [parsePhoneNumber, setParsePhoneNumber] = useState<ParsePhoneNumber>();

  useEffect(() => {
    loadModule().then(({ AsYouType, parsePhoneNumber }) => {
      const asYouType = new AsYouType('US');

      setFormatNumber(asYouType);
      setParsePhoneNumber(() => parsePhoneNumber);
    });
  }, []);

  return { formatNumber, parsePhoneNumber };
};
