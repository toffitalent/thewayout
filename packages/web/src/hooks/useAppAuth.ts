import { useAuth } from '@disruptive-labs/client/react';
import { useRouter } from 'next/router';
import { useCallback, useEffect, useState } from 'react';
import { AppClient, UserType } from '@app/api';

export const useAppAuth = () => useAuth<AppClient>();

export function useAuthRedirect(autoRedirectOnFirstRender = true) {
  const auth = useAppAuth();
  const router = useRouter();
  const [shouldRedirect, setShouldRedirect] = useState(false);

  const redirect = useCallback(() => {
    setShouldRedirect(true);
  }, []);

  // Handle auto-redirect on ready, if enabled
  useEffect(() => {
    if (!router.isReady || auth.isLoading) return;

    if (autoRedirectOnFirstRender && auth.isAuthenticated) redirect();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.isLoading, router.isReady, autoRedirectOnFirstRender, redirect]);

  // Handle manual redirect once auth.user available
  useEffect(() => {
    if (shouldRedirect) {
      setShouldRedirect(false);

      const userType = Object.values(UserType).find((item) => auth.user?.roles?.includes(item));
      router.replace(
        userType ? `/${userType !== UserType.Rsp ? userType : `ssp/case-managers`}/` : '/',
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.user, shouldRedirect]);

  return redirect;
}
