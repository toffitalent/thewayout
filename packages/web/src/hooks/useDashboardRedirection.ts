import { useRouter } from 'next/router';
import { useEffect, useMemo, useState } from 'react';
import { useAppAuth, useAuthRedirect } from './useAppAuth';

export const useDashboardRedirection = () => {
  const router = useRouter();
  const auth = useAppAuth();
  const redirect = useAuthRedirect(false);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    if (auth.isLoading) {
      return;
    }
    if (auth.user && router.asPath === '/') {
      redirect();
    } else {
      setIsLoading(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.user, auth.isLoading, router.asPath]);

  return useMemo(() => isLoading, [isLoading]);
};
