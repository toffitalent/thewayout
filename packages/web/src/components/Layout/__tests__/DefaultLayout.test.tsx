import { useRouter } from 'next/router';
import { useAppAuth } from '@app/hooks';
import { render, waitFor } from '@test';
import { DefaultLayout } from '../DefaultLayout';

jest.mock('@app/hooks', () => ({
  ...jest.requireActual('@app/hooks'),
  useAppAuth: jest.fn(),
}));

describe('DefaultLayout', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockImplementation(() => ({
      events: { on: jest.fn(), off: jest.fn() },
    }));
    (useAppAuth as any).mockImplementation(() => ({
      isAuthenticated: true,
      isLoading: false,
    }));
  });

  test('renders correctly', async () => {
    const { container, findByText } = render(
      <DefaultLayout>
        <div>Page</div>
      </DefaultLayout>,
    );
    await waitFor(async () => expect(await findByText('Dashboard')).toBeInTheDocument());

    expect(container).toMatchSnapshot();
  });
});
