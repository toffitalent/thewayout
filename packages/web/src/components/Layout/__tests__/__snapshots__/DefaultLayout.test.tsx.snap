// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultLayout renders correctly 1`] = `
<div>
  <a
    class="skipNav"
    href="#skip-nav"
  >
    Skip to content
  </a>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <a
        aria-label="Home"
        class="logo"
        href="/client"
      />
      <button
        aria-controls="nav_links"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="toggle"
        type="button"
      >
        <div
          class="hamburger"
        >
          <div
            class="hamburgerInner"
          />
        </div>
      </button>
      <div
        class="wrapper"
      >
        <ul
          class="links"
          id="nav_links"
          role="menu"
        >
          <div
            class="renderedLinks"
          >
            <li
              role="menuitem"
            >
              <a
                class="link"
                href="/client"
              >
                Dashboard
              </a>
            </li>
            <li
              role="menuitem"
            >
              <a
                class="link"
                href="/client/profile"
              >
                My Profile
              </a>
            </li>
            <li
              role="menuitem"
            >
              <a
                class="link"
                href="/client/jobs"
              >
                Jobs
              </a>
            </li>
          </div>
          <li
            class="loggedUserMainMenuItem"
            role="menuitem"
          >
            <button
              class="button lg text c-primary linkButton"
              type="button"
            >
              Account settings
            </button>
          </li>
          <li
            class="loggedUserMainMenuItem"
            role="menuitem"
          >
            <button
              class="button lg text c-primary linkButton"
              type="button"
            >
              Log Out
            </button>
          </li>
        </ul>
      </div>
      <div
        class="loggedUserMenu marginLeft-10"
      >
        <button
          aria-controls="menu-:r0:"
          aria-expanded="false"
          aria-haspopup="menu"
          class="button md text c-primary"
          id="menu-:r0:-trigger"
          role="button"
          type="button"
        >
          <div
            class="avatar bgcolor-dark color-light"
          >
            <div
              aria-label="Test User"
              class="initials"
            >
              TU
            </div>
          </div>
          <span
            aria-hidden="true"
            class="accessoryRight"
          >
            <svg
              aria-hidden="true"
              class="icon menuTriggerAccessory"
              focusable="false"
              stroke-width="2"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                stroke="none"
              />
              <path
                d="M6 9l6 6 6-6"
              />
            </svg>
          </span>
        </button>
        <div
          class="menu"
          data-popper-escaped="true"
          data-popper-placement="bottom-start"
          data-popper-reference-hidden="true"
          style="visibility: hidden; position: absolute; top: 0px; left: 0px; min-width: max-content; will-change: transform, visibility; transform: translate(0px, 8px);"
        >
          <div
            aria-orientation="vertical"
            class="list"
            id="menu-:r0:"
            role="menu"
            tabindex="-1"
          >
            <div
              class="group"
              role="group"
            >
              <button
                class="button md text c-primary item minWidth-40"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Account settings
              </button>
              <button
                class="button md text c-primary item minWidth-40"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Log Out
              </button>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </header>
  <div
    id="skip-nav"
  />
  <main>
    <div>
      Page
    </div>
  </main>
</div>
`;
